### PR TITLE
Decouple device and dtype selections from kornia-vlm models

### LINKED ISSUE #780: [Feature]: Decouple dtype selection from compile-time CUDA feature
### 🚀 Feature Description

The current implementation selects tensor dtype during model initialization by coupling precision choice with compile-time CUDA feature detection. This leads to a static precision policy where:
kornia-rs\kornia-rs\crates\kornia-vlm\src\smolvlm>mod.rs
CUDA → BF16
CPU → F32

This approach does not account for actual runtime hardware capabilities. Some GPUs do not support BF16 efficiently, while others support FP16 but not BF16. As a result, the model may run with suboptimal precision or performance depending on hardware.

This PR proposes introducing runtime-aware dtype selection based on the detected device and its capabilities, instead of compile-time feature flags.

### 📂 Feature Category

Performance Optimization

### 💡 Motivation

The current logic:

- binds dtype choice to compilation features instead of runtime environment
- assumes all CUDA devices support BF16 efficiently
- prevents dynamic optimization for different hardware

reduces portability across heterogeneous systems
Modern inference frameworks determine numeric precision dynamically to maximize throughput while preserving numerical stability.

### 💭 Proposed Solution

Separate device selection from dtype selection and determine precision based on runtime device characteristics.

1. Detect device
2. Query device capabilities
3. Select optimal dtype

### 📚 Library Reference

candle_core::Device
Device::cuda_if_available
CUDA capability query (compute_capability_major())

### 🔄 Alternatives Considered

1. Keep current logic
Rejected because it hardcodes assumptions about hardware capabilities and prevents runtime optimization.

2. User-defined dtype config only
Rejected as sole solution because it shifts responsibility to users and removes automatic optimization.

3. Compile-time feature flags per architecture
Rejected due to poor scalability and inability to adapt to runtime environments.

### 🎯 Use Cases

heterogeneous deployment environments
inference benchmarking
edge devices with older GPUs
mixed CPU/GPU clusters
CI environments without CUDA

### 📝 Additional Context

_No response_

### 🤝 Contribution Intent

- [x] I plan to submit a PR to implement this feature
- [ ] I'm requesting this feature but not planning to implement it

### PR DESCRIPTION
**Pull request by Varun-sai-500**
## 📝 Description

Fixes: #780

## 🛠️ Changes Made

 Incorrect hardware assumption as not all CUDA GPUs support BF16, So I have added checks where it will check with compute compatibility, so it will check before randomly accomodating bf16 to cuda devices

Overall I have added clean code architecture to the device selection and dtype selection 
---

## 🧪 How Was This Tested?
- [x] **Unit Tests:** (List new/updated tests)
- [x] **Manual Verification:** (Describe the steps you took)
- [x] **Performance/Edge Cases:** (How does this handle nulls, large data, etc.?)

---

## 🕵️ AI Usage Disclosure
*Check one of the following:*
- [ ] 🟢 **No AI used.**
- [x] 🟡 **AI-assisted:** I used AI for boilerplate/refactoring but have manually reviewed and tested every line.
- [ ] 🔴 **AI-generated:** (Note: These PRs may be subject to stricter scrutiny or immediate closure if the logic is not explained).

---

## 🚦 Checklist
- [x] I am assigned to the linked issue (required before PR submission)
- [x] The linked issue has been approved by a maintainer
- [x] This PR strictly implements what the linked issue describes (no scope creep)
- [x] I have performed a **self-review** of my code (no "ghost" variables or hallucinations).
- [x] My code follows the existing style guidelines of this project.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [ ] (Optional) I have attached screenshots/recordings for UI changes.

---

## 💭 Additional Context
Add any other context or screenshots about the pull request here.

### CODE REVIEW

**Review by sidd-27 — CHANGES_REQUESTED** [review]
## Review: Incomplete Implementation

While this PR addresses a legitimate architectural issue, the implementation is incomplete and doesn't deliver what's promised.

### ✅ What Works
- Successfully decouples device and dtype selection into separate functions
- Clean separation of concerns

### ❌ Critical Issues

**1. Missing Compute Capability Checks**
The PR description claims to add "checks where it will check with compute compatibility" but this is **not implemented**. The code still only checks `if device.is_cuda()` - the same level of hardware awareness as before.

**2. Undocumented BF16 → F16 Change** 
The code silently switches from `DType::BF16` to `DType::F16` for CUDA devices without explanation. This:
- Isn't mentioned in the PR description  
- May impact model accuracy/performance
- Doesn't solve the original problem (not all GPUs support F16 efficiently either)

**3. No Hardware Capability Detection**
Issue #780 specifically requests runtime capability checks for GPU compatibility. The new code still assumes all CUDA devices support F16.

### Required Changes

The `select_dtype` function needs actual hardware capability detection:

```rust
fn select_dtype(device: &Device) -> DType {
    #[cfg(feature = "cuda")]
    if device.is_cuda() {
        // Check actual compute capability
        if let Ok(capability) = device.compute_capability() {
            if capability >= (8, 0) { // Ampere+ supports BF16
                return DType::BF16;
            } else if capability >= (5, 3) { // Maxwell+ supports F16  
                return DType::F16;
            }
        }
        return DType::F32; // Fallback for older GPUs
    }
    DType::F32
}
```

### Additional Requirements
- Document the BF16 → F16 change and rationale
- Add tests for different GPU capabilities
- Update PR description to match actual implementation

This creates good architectural foundation but needs the core feature - hardware capability detection - to be truly useful.

**Comment by Varun-sai-500** [comment]
@sidd-27  Why these checks need maintainer approval? Can't they run normally as they do in kornia main repo? 

and also, there are similar cases of loose bf16 cases, inside mod.rs of smolvlm2 and also in paligemma, will open a seperate pr for that , but once this gets reviewed, this is a real issue, of bf16 not good for cuda devices which are not latest (ex: ampere)

File: Cargo.toml [hunk 1]
```diff
@@ -54,6 +54,7 @@ kornia-tensor-ops = { path = "crates/kornia-tensor-ops", version = "0.1.11" }
 kornia-vlm = { path = "crates/kornia-vlm", version = "0.1.11" }
 log = "0.4"
 num-traits = "0.2"
+cudarc = "0.11"
```
Comment (sidd-27) [inline Cargo.toml]: this should also be feature gated i think

File: crates/kornia-vlm/src/smolvlm/mod.rs [hunk 2]
```diff
@@ -49,7 +63,16 @@ impl<A: ImageAllocator> SmolVlm<A> {
     pub fn new(config: SmolVlmConfig) -> Result<Self, SmolVlmError> {
         #[cfg(feature = "cuda")]
         let (device, dtype) = match Device::cuda_if_available(0) {
-            Ok(device) => (device, DType::BF16),
+            Ok(device) => {
+                let dtype = if Self::cuda_supports_bf16(0) {
+                    log::info!("GPU supports BF16, using BF16");
+                    DType::BF16
+                } else {
+                    log::warn!("GPU does not support BF16, falling back to FP16");
```
Comment (sidd-27) [inline crates/kornia-vlm/src/smolvlm/mod.rs]: Not a blocker since before this PR those GPUs would just try BF16 and likely fail anyway, so this is strictly better. But worth noting that FP16's narrower exponent range (max ~65504 vs BF16's ~3.4e38) can cause overflow in attention logits. If you've been able to smoke-test on a pre-Ampere GPU, great, otherwise maybe add a log note that FP16 results may not match BF16 quality

Comment (Varun-sai-500) [reply crates/kornia-vlm/src/smolvlm/mod.rs]: I feature gated it now, thanks for the suggestion

**Comment by Varun-sai-500** [comment]
@sidd-27  If there is still anything to update or change, please tell me, I will do it immediately, let's close this today

File: crates/kornia-vlm/src/paligemma/mod.rs [hunk 3]
```diff
@@ -81,6 +81,18 @@ pub struct Paligemma {
 }
 
 impl Paligemma {
+    #[cfg(feature = "cuda")]
+    fn cuda_supports_bf16(device_id: usize) -> bool {
+        use cudarc::driver::{sys::CUdevice_attribute, CudaDevice};
+        if let Ok(dev) = CudaDevice::new(device_id) {
+            if let Ok(major) =
+                dev.attribute(CUdevice_attribute::CU_DEVICE_ATTRIBUTE_COMPUTE_CAPABILITY_MAJOR)
```
Comment (sidd-27) [inline crates/kornia-vlm/src/paligemma/mod.rs]: code being repeated accross files?

**Review by sidd-27 — CHANGES_REQUESTED** [review]

File: crates/kornia-vlm/src/paligemma/mod.rs (hunk 3 above)
Comment (Varun-sai-500) [reply crates/kornia-vlm/src/paligemma/mod.rs]: well actually I implemented the same fix across other files inside mod.rs of various folders such as smolvlm, paligemma, somlvlm2, basically they have the same problem of assuming the bf16 exists in any cuda device, so I implemented my fixes in those files as well, actually I did it as an addition yesterday. Cudarc now is optional in cargo.toml and also implemented well inside cargo of those specific folders as well to support, I think it should be working fine

**Comment by sidd-27** [comment]
what i meant was code deduplication and redundancy, any way to fix that?

**Comment by Varun-sai-500** [comment]
@sidd-27  I just did my changes, can you review now? let's finish it off

File: Cargo.toml (hunk 1 above)
Comment (Varun-sai-500) [reply Cargo.toml]: It's feature gated in my new commit

**Comment by sidd-27** [comment]
the code doesn't compile... i dont think you had locally tested

**Comment by Varun-sai-500** [comment]
Well after de duplication I haven't, I will do it and send it @sidd-27

**Comment by Varun-sai-500** [comment]
@sidd-27  Can you review now? I have tested locally now

**Comment by Varun-sai-500** [comment]
I don't know what's wrong, I will send the fix

File: Cargo.toml [hunk 4]
```diff
@@ -54,6 +54,7 @@ kornia-tensor-ops = { path = "crates/kornia-tensor-ops", version = "0.1.11" }
 kornia-vlm = { path = "crates/kornia-vlm", version = "0.1.11" }
 log = "0.4"
 num-traits = "0.2"
+cudaarc = { version = "0.11"}
```
Comment (sidd-27) [inline Cargo.toml]: theres a typo here thats why it isnt compiling...

**Comment by Varun-sai-500** [comment]
@sidd-27 thanks for suggesting that, now 5 tests are failing and 2 of them are redundant as they are maturin and other type of error which happened on other PRs as well but can you suggest me what else can I do?

**Comment by sidd-27** [comment]
lint still fails, look at contirbution guidelines and agents.md

**Comment by sidd-27** [comment]
> @sidd-27 hello, I have fixed issues, can you just check now, I made sure lint issues are gone, and the code is good now, if there are other errors, they don't belong to the code but some other issues in the repo.

the lint errors are still there, from your changes. Please follow contributor guidelines given in agents.md..

File: crates/kornia-vlm/src/device.rs [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+                    );
+                    DType::F16
+                };
+                (device, dtype)
+            }
+            Err(e) => {
+                log::warn!("CUDA not available, defaulting to CPU: {:?}", e);
+            }
+        }
+    }
```
Comment (sidd-27) [inline crates/kornia-vlm/src/device.rs:24]: also did you compile this code with cuda flags enabled? cause i dont think this complies

**Comment by Varun-sai-500** [comment]
@sidd-27  I ran tests, except for the dataset errors for apriltags (which are not relevant to the PR), all passed!

**Comment by Varun-sai-500** [comment]
@sidd-27 Lint must pass by now

**Comment by Varun-sai-500** [comment]
@sidd-27  Can you answer why?
The same rust lint fails over here, why?

Screenshot 2026-03-30 160151


The same toml error, didn't even need formatting, but it fails here

image

File: crates/kornia-vlm/src/device.rs (hunk 5 above)
Comment (Varun-sai-500) [reply crates/kornia-vlm/src/device.rs:24]: changed code now

**Comment by Varun-sai-500** [comment]
I am also paralleling running lint, let's see

**Comment by Varun-sai-500** [comment]
@sidd-27  CI is green

**Comment by sidd-27** [comment]
lgtm from my end only thing id say is more unit tests, ill let the other maintainers take a look

**Comment by Varun-sai-500** [comment]
@sidd-27 Can you approve the PR and ping other maintainers ?

**Review by sidd-27 — APPROVED** [review]

**Comment by Varun-sai-500** [comment]
@ducha-aiki  Can you review this approved PR? and merge? Thanks

**Comment by sidd-27** [comment]
sorry for the delay i was a bit busy

**Comment by Varun-sai-500** [comment]
Thank you brother, had this listed on my resume, had to prove that it's been merged, I am too in pre-final yr @sidd-27
